# Assignment: Preparing Data for Analysis (Modified Titanic)

![](https://github.com/kaopanboonyuen/2110446_DataScience_2021s2/raw/main/%20files/hw.png)

In [592]:
#Import the libraries
import numpy as np
import pandas as pd

In [593]:
df = pd.read_csv('https://github.com/kaopanboonyuen/2110446_DataScience_2021s2/raw/main/datasets/hw/titanic_training_dataset_v2.csv')

In [594]:
df.head()

PassengerId  Survived  Pclass  \
0            1       0.0     3.0   
1            2       1.0     1.0   
2            3       1.0     3.0   
3            4       1.0     1.0   
4            5       0.0     3.0   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0    1.0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0    1.0   
2                             Heikkinen, Miss. Laina  female  26.0    0.0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
4                           Allen, Mr. William Henry    male  35.0    0.0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [595]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     865 non-null    float64
 2   Pclass       828 non-null    float64
 3   Name         866 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        858 non-null    float64
 7   Parch        891 non-null    int64  
 8   Ticket       852 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     803 non-null    object 
dtypes: float64(5), int64(2), object(5)
memory usage: 83.7+ KB


In [596]:
df.describe(include='all')

PassengerId    Survived      Pclass                     Name   Sex  \
count    891.000000  865.000000  828.000000                      866   891   
unique          NaN         NaN         NaN                      866     2   
top             NaN         NaN         NaN  Braund, Mr. Owen Harris  male   
freq            NaN         NaN         NaN                        1   577   
mean     446.000000    0.383815    2.304348                      NaN   NaN   
std      257.353842    0.486595    0.836952                      NaN   NaN   
min        1.000000    0.000000    1.000000                      NaN   NaN   
25%      223.500000    0.000000    2.000000                      NaN   NaN   
50%      446.000000    0.000000    3.000000                      NaN   NaN   
75%      668.500000    1.000000    3.000000                      NaN   NaN   
max      891.000000    1.000000    3.000000                      NaN   NaN   

               Age       SibSp       Parch Ticket        Fare    Cabin  \
count   714.000000  858.000000  891.000000    852  891.000000      204   
unique         NaN         NaN         NaN    653         NaN      147   
top            NaN         NaN         NaN   1601         NaN  B96 B98   
freq           NaN         NaN         NaN      7         NaN        4   
mean     29.699118    0.524476    0.381594    NaN   32.204208      NaN   
std      14.526497    1.115284    0.806057    NaN   49.693429      NaN   
min       0.420000    0.000000    0.000000    NaN    0.000000      NaN   
25%      20.125000    0.000000    0.000000    NaN    7.910400      NaN   
50%      28.000000    0.000000    0.000000    NaN   14.454200      NaN   
75%      38.000000    1.000000    0.000000    NaN   31.000000      NaN   
max      80.000000    8.000000    6.000000    NaN  512.329200      NaN   

       Embarked  
count       803  
unique        3  
top           S  
freq        580  
mean        NaN  
std         NaN  
min         NaN  
25%         NaN  
50%         NaN  
75%         NaN  
max         NaN

# 1) Load data & review the data

<font color='blue'>Q1) What is the shape of this dataset? (rows & columns)</font>

ans: (891, 12)

In [597]:
df.shape

(891, 12)

# 2) Drop unqualified variables

*   Drop variables with missing > 50%
*   Drop categorical variables with flat values > 70% (variables with the same value in the same column)

<font color='blue'>Q2) How many columns do we have left?</font>

ans: 10 columns


In [598]:
half_count = (len(df) + 1) // 2
df = df.dropna(thresh=half_count, axis=1) 
df.shape



(891, 11)

In [599]:
df.head(5)

PassengerId  Survived  Pclass  \
0            1       0.0     3.0   
1            2       1.0     1.0   
2            3       1.0     3.0   
3            4       1.0     1.0   
4            5       0.0     3.0   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0    1.0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0    1.0   
2                             Heikkinen, Miss. Laina  female  26.0    0.0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
4                           Allen, Mr. William Henry    male  35.0    0.0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

In [600]:
object_cols = df.select_dtypes(include=['object']).columns
object_cols

Index(['Name', 'Sex', 'Ticket', 'Embarked'], dtype='object')

In [601]:
cate_cols = ['PassengerId',	'Survived',	'Pclass', 'Name',	'Sex', 'SibSp',	'Parch', 'Ticket','Embarked'] 
df[cate_cols]

PassengerId  Survived  Pclass  \
0              1       0.0     3.0   
1              2       1.0     1.0   
2              3       1.0     3.0   
3              4       1.0     1.0   
4              5       0.0     3.0   
..           ...       ...     ...   
886          887       0.0     2.0   
887          888       1.0     1.0   
888          889       0.0     3.0   
889          890       1.0     1.0   
890          891       0.0     3.0   

                                                  Name     Sex  SibSp  Parch  \
0                              Braund, Mr. Owen Harris    male    1.0      0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female    1.0      0   
2                               Heikkinen, Miss. Laina  female    0.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female    1.0      0   
4                             Allen, Mr. William Henry    male    0.0      0   
..                                                 ...     ...    ...    ...   
886                              Montvila, Rev. Juozas    male    0.0      0   
887                       Graham, Miss. Margaret Edith  female    0.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female    1.0      2   
889                              Behr, Mr. Karl Howell    male    0.0      0   
890                                Dooley, Mr. Patrick    male    0.0      0   

               Ticket Embarked  
0           A/5 21171        S  
1            PC 17599        C  
2    STON/O2. 3101282        S  
3              113803        S  
4              373450        S  
..                ...      ...  
886            211536        S  
887            112053        S  
888        W./C. 6607        S  
889            111369        C  
890            370376        Q  

[891 rows x 9 columns]

In [602]:
percent_same_value = df[cate_cols].apply(lambda x: x.value_counts(sort = True, ascending = False).values[0]/df.shape[0]*100)
percent_same_value



PassengerId     0.112233
Survived       59.820426
Pclass         50.953984
Name            0.112233
Sex            64.758698
SibSp          65.993266
Parch          76.094276
Ticket          0.785634
Embarked       65.095398
dtype: float64

In [603]:
columns_to_drop = percent_same_value[percent_same_value > 70].index
columns_to_drop

Index(['Parch'], dtype='object')

In [604]:
df = df.drop(columns_to_drop, axis=1)
df

PassengerId  Survived  Pclass  \
0              1       0.0     3.0   
1              2       1.0     1.0   
2              3       1.0     3.0   
3              4       1.0     1.0   
4              5       0.0     3.0   
..           ...       ...     ...   
886          887       0.0     2.0   
887          888       1.0     1.0   
888          889       0.0     3.0   
889          890       1.0     1.0   
890          891       0.0     3.0   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0    1.0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0    1.0   
2                               Heikkinen, Miss. Laina  female  26.0    0.0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
4                             Allen, Mr. William Henry    male  35.0    0.0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0    0.0   
887                       Graham, Miss. Margaret Edith  female  19.0    0.0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN    1.0   
889                              Behr, Mr. Karl Howell    male  26.0    0.0   
890                                Dooley, Mr. Patrick    male  32.0    0.0   

               Ticket     Fare Embarked  
0           A/5 21171   7.2500        S  
1            PC 17599  71.2833        C  
2    STON/O2. 3101282   7.9250        S  
3              113803  53.1000        S  
4              373450   8.0500        S  
..                ...      ...      ...  
886            211536  13.0000        S  
887            112053  30.0000        S  
888        W./C. 6607  23.4500        S  
889            111369  30.0000        C  
890            370376   7.7500        Q  

[891 rows x 10 columns]

In [605]:
df.shape

(891, 10)

# 3) Remove all rows with missing target (the variable "Survived")

<font color='blue'>Q3) How many rows do we have left?</font>

ans: 865 rows

In [606]:
df = df[df['Survived'].notna()]
df.shape

(865, 10)

# 4) Handle outliers 

For the variable “Fare”, replace outlier values with the boundary values


*   If value < (Q1 - 1.5IQR), relace with (Q1 - 1.5IQR)
*   If value > (Q3 + 1.5IQR), relace with (Q3 + 1.5IQR)

<font color='blue'>Q4) What is the mean of “Fare” after replacing the outliers (round 2 decimal points)?</font>
 
ans: 24.04


In [607]:
q25 = df["Fare"].quantile(0.25)
q75 = df["Fare"].quantile(0.75)
iqr = q75 - q25

min = q25 - (iqr*1.5)
max = q75 + (iqr*1.5)
print('Min: ', min)
print('Max: ', max)

Min:  -26.7605
Max:  65.6563


In [608]:
df["Fare"] = df["Fare"].where(df["Fare"] >= min, min)
df["Fare"] = df["Fare"].where(df["Fare"] <= max, max)


<ipython-input-608-6d7832530d8f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Fare"] = df["Fare"].where(df["Fare"] >= min, min)
<ipython-input-608-6d7832530d8f>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Fare"] = df["Fare"].where(df["Fare"] <= max, max)


In [609]:
df['Fare'].mean()

24.038996878612718

# 5) Impute missing value



*   Show the number of missing values in each variable
*   Impute missing values with mean & mode
*   Show the number of missing values again after missing value imputation

<font color='blue'>Q5) Which variable has the largest number of missing values?</font>

ans: Age (172)



In [610]:
has_nan_list = []
for var in df.columns.tolist():
    missing_values = df[var].isna().sum()
    print(var, missing_values)
    if missing_values > 0: has_nan_list.append(var)
has_nan_list

PassengerId 0
Survived 0
Pclass 63
Name 25
Sex 0
Age 172
SibSp 33
Ticket 39
Fare 0
Embarked 87


['Pclass', 'Name', 'Age', 'SibSp', 'Ticket', 'Embarked']

In [611]:
for col in has_nan_list:
    if col in cate_cols:
        df[col] = df[col].fillna(df[col].mode().iloc[0])
    else: df[col] = df[col].fillna(df[col].mean())

<ipython-input-611-69838a1115a0>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col] = df[col].fillna(df[col].mode().iloc[0])
<ipython-input-611-69838a1115a0>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  else: df[col] = df[col].fillna(df[col].mean())


In [612]:
for var in df.columns.tolist():
    missing_values = df[var].isna().sum()
    print(var, missing_values)

PassengerId 0
Survived 0
Pclass 0
Name 0
Sex 0
Age 0
SibSp 0
Ticket 0
Fare 0
Embarked 0


# 6) Convert categorical to numeric values



*   Drop the variables “Name” & “Ticket”
*   For the variables “Sex” & “Embarked”, perform the dummy coding and drop the first level. Also, drop those original variables (“Sex” & “Embarked”)

<font color='blue'>Q6) How many columns do we have?</font>

ans: 9 columns


In [613]:
df = df.drop(['Name', 'Ticket'], axis = 1)
df

PassengerId  Survived  Pclass     Sex        Age  SibSp     Fare Embarked
0              1       0.0     3.0    male  22.000000    1.0   7.2500        S
1              2       1.0     1.0  female  38.000000    1.0  65.6563        C
2              3       1.0     3.0  female  26.000000    0.0   7.9250        S
3              4       1.0     1.0  female  35.000000    1.0  53.1000        S
4              5       0.0     3.0    male  35.000000    0.0   8.0500        S
..           ...       ...     ...     ...        ...    ...      ...      ...
886          887       0.0     2.0    male  27.000000    0.0  13.0000        S
887          888       1.0     1.0  female  19.000000    0.0  30.0000        S
888          889       0.0     3.0  female  29.892266    1.0  23.4500        S
889          890       1.0     1.0    male  26.000000    0.0  30.0000        C
890          891       0.0     3.0    male  32.000000    0.0   7.7500        Q

[865 rows x 8 columns]

In [614]:
dummies = pd.get_dummies(df[["Sex", "Embarked"]], drop_first=True)
df = pd.concat([df, dummies], axis=1)
df


PassengerId  Survived  Pclass     Sex        Age  SibSp     Fare  \
0              1       0.0     3.0    male  22.000000    1.0   7.2500   
1              2       1.0     1.0  female  38.000000    1.0  65.6563   
2              3       1.0     3.0  female  26.000000    0.0   7.9250   
3              4       1.0     1.0  female  35.000000    1.0  53.1000   
4              5       0.0     3.0    male  35.000000    0.0   8.0500   
..           ...       ...     ...     ...        ...    ...      ...   
886          887       0.0     2.0    male  27.000000    0.0  13.0000   
887          888       1.0     1.0  female  19.000000    0.0  30.0000   
888          889       0.0     3.0  female  29.892266    1.0  23.4500   
889          890       1.0     1.0    male  26.000000    0.0  30.0000   
890          891       0.0     3.0    male  32.000000    0.0   7.7500   

    Embarked  Sex_male  Embarked_Q  Embarked_S  
0          S         1           0           1  
1          C         0           0           0  
2          S         0           0           1  
3          S         0           0           1  
4          S         1           0           1  
..       ...       ...         ...         ...  
886        S         1           0           1  
887        S         0           0           1  
888        S         0           0           1  
889        C         1           0           0  
890        Q         1           1           0  

[865 rows x 11 columns]

In [615]:
df = df.drop(["Sex", "Embarked"], axis=1)
df

PassengerId  Survived  Pclass        Age  SibSp     Fare  Sex_male  \
0              1       0.0     3.0  22.000000    1.0   7.2500         1   
1              2       1.0     1.0  38.000000    1.0  65.6563         0   
2              3       1.0     3.0  26.000000    0.0   7.9250         0   
3              4       1.0     1.0  35.000000    1.0  53.1000         0   
4              5       0.0     3.0  35.000000    0.0   8.0500         1   
..           ...       ...     ...        ...    ...      ...       ...   
886          887       0.0     2.0  27.000000    0.0  13.0000         1   
887          888       1.0     1.0  19.000000    0.0  30.0000         0   
888          889       0.0     3.0  29.892266    1.0  23.4500         0   
889          890       1.0     1.0  26.000000    0.0  30.0000         1   
890          891       0.0     3.0  32.000000    0.0   7.7500         1   

     Embarked_Q  Embarked_S  
0             0           1  
1             0           0  
2             0           1  
3             0           1  
4             0           1  
..          ...         ...  
886           0           1  
887           0           1  
888           0           1  
889           0           0  
890           1           0  

[865 rows x 9 columns]

In [616]:
df.shape

(865, 9)

# 7) Partition data



*   Split train/test split with stratification using 70%:30% and random seed with 12345
*   Show a proportion between survived (1) and died (0) in all data sets (total data, train, test)

<font color='blue'>Q7) What is a proportion between survived (1) and died (0) in the training data?</font>

On train data

dead = 0.6165289256198347

survived = 0.3834710743801653


In [617]:
from sklearn.model_selection import train_test_split

X = df
y = df['Survived']

X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.3, random_state=12345)

In [618]:
survived_all = (df['Survived']==1).sum()
died_all = (df['Survived']==0).sum()

print('On total data')
print('dead =',died_all/df['Survived'].shape[0])
print('survived =',survived_all/df['Survived'].shape[0])

On total data
dead = 0.6161849710982659
survived = 0.3838150289017341


In [619]:
survived_test = (y_test==1).sum()
died_test = (y_test==0).sum()

print('On test data')
print('dead =',died_test/y_test.shape[0])
print('survived =',survived_test/y_test.shape[0])

On test data
dead = 0.6153846153846154
survived = 0.38461538461538464


In [620]:
survived_train = (y_train==1).sum()
died_train = (y_train==0).sum()

print('On train data')
print('dead =',died_train/y_train.shape[0])
print('survived =',survived_train/y_train.shape[0])

On train data
dead = 0.6165289256198347
survived = 0.3834710743801653


In [621]:
print('Number of rows in training set =',y_train.shape[0])
print('Number of rows in testing set =',y_test.shape[0])

Number of rows in training set = 605
Number of rows in testing set = 260
